src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+            _writingHead = segment;
+
+            return segment;
+        }
+
+        private BufferSegment CreateSegmentUnsynchronized()
+        {
+            if (_pooledSegmentCount > 0)
+            {
+                _pooledSegmentCount--;
+                return _bufferSegmentPool[_pooledSegmentCount];
+            }
+
+            return new BufferSegment();
+        }
+
+        private void ReturnSegmentUnsynchronized(BufferSegment segment)
+        {
+            if (_pooledSegmentCount < _bufferSegmentPool.Length)
+            {
+                _bufferSegmentPool[_pooledSegmentCount] = segment;
+                _pooledSegmentCount++;
+            }
+        }
+
+        internal void Commit()
+        {
+            // Changing commit head shared with Reader
+            lock (_sync)
+            {
+                CommitUnsynchronized();
+            }
+        }
+
+        internal void CommitUnsynchronized()
+        {
+            if (_writingHead == null)
+            {
+                // Nothing written to commit
+                return;
+            }
+
+            if (_readHead == null)
+            {
+                // Update the head to point to the head of the buffer.
+                // This happens if we called alloc(0) then write
+                _readHead = _commitHead;
+                _readHeadIndex = 0;
+            }
+
+            // Always move the commit head to the write head
+            _commitHead = _writingHead;
+            _commitHeadIndex = _writingHead.End;
+            _length += _currentWriteLength;
+
+            // Do not reset if reader is complete
+            if (_maximumSizeHigh > 0 &&
+                _length >= _maximumSizeHigh &&
+                !_readerCompletion.IsCompleted)
+            {
+                _writerAwaitable.Reset();
+            }
+
+            // Clear the writing state
+            _writingHead = null;
+            _currentWriteLength = 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void Advance(int bytesWritten)
+        {
+            if (_writingHead == null)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NotWritingNoAlloc();
+            }
+
+            if (bytesWritten > 0)
+            {
+                Debug.Assert(!_writingHead.ReadOnly);
+                Debug.Assert(_writingHead.Next == null);
+
+                Memory<byte> buffer = _writingHead.AvailableMemory;
+                int bufferIndex = _writingHead.End + bytesWritten;
+
+                if (bufferIndex > buffer.Length)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_AdvancingPastBufferSize();
+                }
+
+                _writingHead.End = bufferIndex;
+                _currentWriteLength += bytesWritten;
+            }
+            else if (bytesWritten < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.bytesWritten);
```

Comment:
Since this is an internal method, should this be a Debug Assert instead of a throw? We could do this check at the call-site instead.

Reply:
It is called by public method directly.